#### Swapnil Chavan
#### DSC 550
#### Week 9 : Hyperparameter Tuning
#### 9.2 Exercise: Best Model Selection and Hyperparameter Tuning


In this exercise, you will work with the Loan_Train.csv dataset which can be downloaded from this link: Loan Approval Data Set. 



1. Import the dataset and ensure that it loaded properly

In [153]:
import pandas as pd

In [166]:
df = pd.read_csv("C:/Users/swapn/Data Mining/Loan_Train.csv")

In [167]:
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

2. Prepare the data for modeling by performing the following steps:
* Drop the column “Load_ID.”
* Drop any rows with missing data.
* Convert the categorical features into dummy variables.


In [168]:
df.drop(columns=['Loan_ID'], inplace=True)

In [169]:
df.dropna(inplace=True)

In [170]:
df2 = pd.get_dummies(df, columns=["Gender","Married","Self_Employed","Property_Area", "Education", "Dependents"])

In [171]:
df2.head()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   

   Credit_History Loan_Status  Gender_Female  Gender_Male  Married_No  \
1             1.0           N          False         True       False   
2             1.0           Y          False         True       False   
3             1.0           Y          False         True       False   
4             1.0           Y          False         True        True   
5             1.0           Y          False         True       False   

   Married_Yes  Self_Employed_No  Self_Employed_Yes  Property_Area_Rural  \
1         True              True              False                 True   
2         True             False               True                False   
3         True              True              False                False   
4        False              True              False                False   
5         True             False               True                False   

   Property_Area_Semiurban  Property_Area_Urban  Education_Graduate  \
1                    False                False                True   
2                    False                 True                True   
3                    False                 True               False   
4                    False                 True                True   
5                    False                 True                True   

   Education_Not Graduate  Dependents_0  Dependents_1  Dependents_2  \
1                   False         False          True         False   
2                   False          True         False         False   
3                    True          True         False         False   
4                   False          True         False         False   
5                   False         False         False          True   

   Dependents_3+  
1          False  
2          False  
3          False  
4          False  
5          False

3. Split the data into a training and test set, where the “Loan_Status” column is the target.

In [162]:
from sklearn.model_selection import train_test_split

In [172]:
df2.head()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   

   Credit_History Loan_Status  Gender_Female  Gender_Male  Married_No  \
1             1.0           N          False         True       False   
2             1.0           Y          False         True       False   
3             1.0           Y          False         True       False   
4             1.0           Y          False         True        True   
5             1.0           Y          False         True       False   

   Married_Yes  Self_Employed_No  Self_Employed_Yes  Property_Area_Rural  \
1         True              True              False                 True   
2         True             False               True                False   
3         True              True              False                False   
4        False              True              False                False   
5         True             False               True                False   

   Property_Area_Semiurban  Property_Area_Urban  Education_Graduate  \
1                    False                False                True   
2                    False                 True                True   
3                    False                 True               False   
4                    False                 True                True   
5                    False                 True                True   

   Education_Not Graduate  Dependents_0  Dependents_1  Dependents_2  \
1                   False         False          True         False   
2                   False          True         False         False   
3                    True          True         False         False   
4                   False          True         False         False   
5                   False         False         False          True   

   Dependents_3+  
1          False  
2          False  
3          False  
4          False  
5          False

In [1]:
y = df2['Loan_Status']
X = df2.drop('Loan_Status', axis=1)


# Split into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

NameError: name 'df2' is not defined

4. Create a pipeline with a min-max scaler and a KNN classifier (see section 15.3 in the Machine Learning with Python Cookbook).

Now, you will create a pipeline that consists of a min-max scaler and a KNN classifier. The min-max scaler scales the features to a specified range, and the KNN classifier is a machine learning algorithm for classification tasks.

In [174]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier

# Create the pipeline
pipeline = Pipeline([
    ('scaler', MinMaxScaler()),
    ('classifier', KNeighborsClassifier())
])

The Pipeline class from scikit-learn is used to create a pipeline object. The pipeline consists of two steps: a min-max scaler named 'scaler' and a KNN classifier named 'classifier'. The scaler will be applied to the features before feeding them into the classifier.

5. Fit a default KNN classifier to the data with this pipeline. Report the model accuracy on the test set. Note: Fitting a pipeline model works just like fitting a regular model.

In [175]:
# Fit the default KNN classifier to the data
pipeline.fit(X_train, y_train)

# Evaluate the model on the test set
accuracy = pipeline.score(X_test, y_test)
print("The accuracy of the default KNN classifier on the test set is: {:.2f}".format(accuracy))

The accuracy of the default KNN classifier on the test set is: 0.74


6. Create a search space for your KNN classifier where your “n_neighbors” parameter varies from 1 to 10. (see section 15.3 in the Machine Learning with Python Cookbook).

Let's create a seatch space for KNN classifier, using GridSearchCV for hyperparameter tuning

In [177]:
from sklearn.model_selection import GridSearchCV
sp = {'classifier__n_neighbors': range(1, 11)}

7. Fit a grid search with your pipeline, search space, and 5-fold cross-validation to find the best value for the “n_neighbors” parameter.

Let's fit a grid search with the pipeline, search space, and 5-fold cross-validation and find our "n_neighbors".





In [179]:
gs = GridSearchCV(pipeline, sp, cv=5)
gs.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', MinMaxScaler()),
                                       ('classifier', KNeighborsClassifier())]),
             param_grid={'classifier__n_neighbors': range(1, 11)})

8. Finding the accuracy of the best model from grid search on the test set

find the accuracy of the best model obtained from the grid search on the test set.

In [ ]:
bm = gs.best_estimator_
accuracy = best_model.score(X_test, y_test)

In [180]:
print("the best model from grid search on the test set: {:.2f}".format(accuracy))

the best model from grid search on the test set: 0.74


9.Expanding the search space with logistic regression and random forest models
In this step, you will repeat the process of defining a search space and fitting a grid search, but this time including logistic regression and random forest models in addition to the KNN classifier.



In [202]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
search_space = {
    'classifier': [KNeighborsClassifier(), LogisticRegression(), RandomForestClassifier()],
}
grid_search = GridSearchCV(pipeline, search_space, cv=5)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', MinMaxScaler()),
                                       ('classifier', KNeighborsClassifier())]),
             param_grid={'classifier': [KNeighborsClassifier(),
                                        LogisticRegression(),
                                        RandomForestClassifier()]})

In the expanded search space, you include the logistic regression and random forest models as additional options for the "classifier" hyperparameter. By specifying multiple classifiers and their corresponding hyperparameters, the grid search will consider all possible combinations and identify the best model and hyperparameters based on cross-validated performance.




 Finding the best model and hyperparameters from the grid search

In [203]:

best_model = grid_search.best_estimator_
best_params = grid_search.best_params_
accuracy = best_model.score(X_test, y_test)
print("Accuracy of the best model from grid search on the test set: {:.2f}".format(accuracy))
print("Best model and hyperparameters found from grid search:")
print(best_model)
print(best_params)

Accuracy of the best model from grid search on the test set: 0.80
Best model and hyperparameters found from grid search:
Pipeline(steps=[('scaler', MinMaxScaler()),
                ('classifier', LogisticRegression())])
{'classifier': LogisticRegression()}


11. Summary

In summary, by using a pipeline and grid search, we performed hyperparameter tuning and found the best model to be a LogisticRegression with the specified hyperparameters. This model achieved an accuracy of 0.80 on the test set, indicating its good performance in predicting loan approval status based on the provided dataset.
